millw

```diff
@@ -106,34 +106,53 @@ fi
 MILL="${MILL_DOWNLOAD_PATH}/${MILL_VERSION}"
 
 try_to_use_system_mill() {
+  if [ "$(uname)" != "Linux" ]; then
+    return 0
+  fi
+
   MILL_IN_PATH="$(command -v mill || true)"
 
   if [ -z "${MILL_IN_PATH}" ]; then
-    return
+    return 0
   fi
 
-  UNIVERSAL_SCRIPT_MAGIC="@ 2>/dev/null # 2>nul & echo off & goto BOF"
+  SYSTEM_MILL_PATH=$(readlink -e "${MILL_IN_PATH}")
+  SYSTEM_MILL_SIZE=$(stat --format=%s "${SYSTEM_MILL_PATH}")
+  SYSTEM_MILL_MTIME=$(stat --format=%y "${SYSTEM_MILL_PATH}")
 
-  if ! head -c 128 "${MILL_IN_PATH}" | grep -qF "${UNIVERSAL_SCRIPT_MAGIC}"; then
-    if [ -n "${MILLW_VERBOSE}" ]; then
-      echo "Could not determine mill version of ${MILL_IN_PATH}, as it does not start with the universal script magic2" 1>&2
+  MILL_USER_CACHE_DIR="${XDG_CACHE_HOME:-${HOME}/.cache}"/mill
+  if [ ! -d "${MILL_USER_CACHE_DIR}" ]; then
+    mkdir -p "${MILL_USER_CACHE_DIR}"
+  fi
+
+  SYSTEM_MILL_INFO_FILE="${MILL_USER_CACHE_DIR}"/system-mill-info
+  if [ -f "${SYSTEM_MILL_INFO_FILE}" ]; then
+    # shellcheck source=/dev/null
+    . "${SYSTEM_MILL_INFO_FILE}"
```

Comment:
Thanks for your feedback.

Bash's parsing abilities are limited, even more the one of POSIX sh. Hence, sourcing configuration or state files is a common idiom in scripts for both. But not only my shell, also my window manager and my editor source files from my home directory on startup. 

And I do not see the problem this is supposed to solve. What is the scenario that you have in mind that causes unforeseen side-effects? Security? As soon as someone, e.g., a malicious entity, is able to temper with your `$HOME` it is already game over. `millw` not sourcing a file under `$HOME` does not make a difference. Robustness? If the state file gets corrupt due I/O errors, then it is likely that other files under the user's $HOME are affected too. Plus parsing instead of sourcing does not help much in this case either.

As it is right now, I am inclined to not put effort into this, as I am believe the expense for doing so is not justified. You are, of course, entitled to reject the PR. Especially since it is probably a somewhat niche feature.

However, maybe I am missing something. If you are willing to elaborate a bit more on the scenario you have in mind, then my assessment may change.

Reply:
Meh, I somehow got bored and ended up doing it anyway. The resulting code is hard to read and really makes me feel bad…

Reply:
Thanks for the changes!

Those shell scripts you use in your system are either written by you or installed (and reviewed) by your trusted package manager. But beside any security considerations, as the maintainer of the script, I simply don't like to debug scripts which are hard to reason about.

Reply:
> Those shell scripts you use in your system are either written by you or installed (and reviewed) by your trusted package manager.

And this also applies to the `system-mill-info` info.

It appears a bit irrational to consider `milw` sourcing `system-mill-info` problematic, while `millw` executing the `mill` downloaded right next to `system-mill-info` being unproblematic.

> But beside any security considerations, as the maintainer of the script, I simply don't like to debug scripts which are hard to reason about.

You find four lines of
```bash
CACHED_SYSTEM_MILL_PATH=$(sed -n '1p' "${SYSTEM_MILL_INFO_FILE}" | cut -d= -f2 | sed 's/"\(.*\)"/\1/')
```
to be easier to reason about than
```bash
. "${SYSTEM_MILL_INFO_FILE}"
```
?

I hope you can someday explain this to me in a relaxed atmosphere with beverages of our choice in your hand. :)

Reply:
> > Those shell scripts you use in your system are either written by you or installed (and reviewed) by your trusted package manager.
> 
> And this also applies to the `system-mill-info` info.
> 
> It appears a bit irrational to consider `milw` sourcing `system-mill-info` problematic, while `millw` executing the `mill` downloaded right next to `system-mill-info` being unproblematic.

This is the main purpose of the script, it's documented and the expected main effect.

> > But beside any security considerations, as the maintainer of the script, I simply don't like to debug scripts which are hard to reason about.
> 
> You find four lines of
> 
> ```shell
> CACHED_SYSTEM_MILL_PATH=$(sed -n '1p' "${SYSTEM_MILL_INFO_FILE}" | cut -d= -f2 | sed 's/"\(.*\)"/\1/')
> ```
> 
> to be easier to reason about than
> 
> ```shell
> . "${SYSTEM_MILL_INFO_FILE}"
> ```

Yes. TBH, I'd like to either have some comment on it or have it in a dedicated function with a good name. But beside, this line doesn't magically fill env variables but explicitly sets the value of a explicitly named variable with the outcome of the rhs expression.

> ? Needless to say that the "parsing" is not robust as it fails on all kinds of corner 
cases, like `FOO="bar\"baz"`.

Yeah, there is a chance to produce a false cache-miss. I think the whole feature is per se  questionable and creates more friction than anticipated. This PR fixed an existing feature, hence I merged it. But TBH, I'm pretty sure, if someone want's to use her locally installed `mill`, she will just run it directly. Maybe, we should consider removing it altogether, if you think it's not robust enough?

Reply:
> creates more friction than anticipated

Do you feel the feature itself creates friction or the discussion around how it should be implemented?


> Yes. TBH, I'd like to either have some comment on it or have it in a dedicated function with a good name

I had the same thought. The switch away from sourcing the file was a quick hack to demonstrate how it could be done. I'll submit a PR that creates an extra function and adds some code comments.


> But TBH, I'm pretty sure, if someone want's to use her locally installed mill, she will just run it directly

Directly running the system `mill` version is not the equivalent action to what the "use system mill if possible" feature provides. But I think you know that.


> Maybe, we should consider removing it altogether, if you think it's not robust enough?

I never stated that I believe that the feature is not robust enough. I argue that the switching away from sourcing the file is not sensible.

And I am sure you can do better when arguing than subliminally threatening to remove a feature.